packages/yew/src/dom_bundle/braw.rs

```diff
@@ -0,0 +1,220 @@
+use wasm_bindgen::JsCast;
+use web_sys::Element;
+
+use crate::dom_bundle::bnode::BNode;
+use crate::dom_bundle::traits::{Reconcilable, ReconcileTarget};
+use crate::dom_bundle::utils::insert_node;
+use crate::dom_bundle::BSubtree;
+use crate::html::AnyScope;
+use crate::virtual_dom::VRaw;
+use crate::NodeRef;
+
+#[derive(Debug)]
+pub struct BRaw {
+    reference: NodeRef,
+    children_count: usize,
+}
+
+impl BRaw {
+    fn create_elements(html: &str) -> Vec<Element> {
+        let div = gloo::utils::document().create_element("div").unwrap();
+        let html = html.trim();
+        div.set_inner_html(html);
+        let children = div.children();
+        let children = js_sys::Array::from(&children);
+        let children = children.to_vec();
+        children
+            .into_iter()
+            .map(|it| it.unchecked_into())
+            .collect::<Vec<_>>()
+    }
+
+    fn detach_bundle(&self, parent: &Element) {
+        if let Some(node) = self.reference.cast::<Element>() {
+            let mut next_sibling = node.unchecked_ref::<Element>().next_sibling();
+
+            parent
+                .remove_child(&node)
+                .expect("failed to remove braw node");
+            let len = self.children_count - 1;
```

Comment:
This underflows when `children_count == 0`. I think `children_count > 0` implies `self.reference.get() == Some(_)` but not the other way around. That is, the easiest way I can see this working is

```rust
let mut next_node = self.reference.get();

for _ in 0..self.children_count {
    let node = next_node.expect("there should be another node"); // Maybe just bail in non-debug mode instead of panic-ing
    parent.remove_child(&node).expect("failed to remove braw node");
    next_node = node.next_sibling();
}
```

Reply:
> There must be at least 1 children

I don't see how that is true, did I miss an assertion for that? In my mind, `Html::from_raw_html("")`, mounted at the end of a container, will produce zero children, with `self.reference.get()` returning `None`.

Reply:
The loop only gets executed if `self.reference.get()` returns `Some(_)`. There's no need to do any computation if nothing was mounted

Reply:
That's kind of my point: If nothing gets mounted, `self.reference` still gets set to `next_sibling` so that the position is saved (that in itself is not the issue though). I think of `self.reference` as like a cursor pointing to the start of the range of inserted elements. If there were some children that were inserted, then we get `get()` the first of those, and then iterate forward with `.next_sibling()`. But if no children were inserted, then the "cursor" still points to the `next_sibling` and could return `Some(_)` if such a sibling exists.

The conclusion is that `self.reference.get()` returning `Some(_)` can also happen when `self.children_count == 0` and nothing was inserted. And that's the bug, I believe.

Reply:
It makes sense but I can't come up with a test case to reproduce it. It would be great if you provide something to test this with

Reply:
Nice. For completeness, I think a resilient test case against this could pass in some `NodeRef::new()`, like with the following diff - a similar change could be done for the other `detach*` test cases. That should also fail with the previous version.

```diff
diff --git a/packages/yew/src/dom_bundle/braw.rs b/packages/yew/src/dom_bundle/braw.rs
index 8b8b4a4b..72efd102 100644
--- a/packages/yew/src/dom_bundle/braw.rs
+++ b/packages/yew/src/dom_bundle/braw.rs
@@ -218,14 +218,16 @@ mod tests {
     #[test]
     fn braw_detach_works_empty() {
         let (root, scope, parent) = setup_parent();
+        let end = gloo::utils::document().create_text_node("END");
+        parent.append_child(&end).unwrap();
 
         const HTML: &str = "";
         let elem = VNode::from_raw_html(HTML.into());
-        let (_, mut elem) = elem.attach(&root, &scope, &parent, NodeRef::default());
+        let (_, mut elem) = elem.attach(&root, &scope, &parent, NodeRef::new(end.into()));
         assert_braw(&mut elem);
-        assert_eq!(parent.inner_html(), HTML);
+        assert_eq!(parent.inner_html(), format!("{}END", HTML));
         elem.detach(&root, &parent, false);
-        assert_eq!(parent.inner_html(), "");
+        assert_eq!(parent.inner_html(), "END");
     }
 
     fn assert_braw(node: &mut BNode) -> &mut BRaw {

```

Reply:
I added those tests